src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -243,19 +251,26 @@ public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescrip
                         continue;
                     }
 
+
                     _notFullyQualifiedComponents ??= new();
                     _notFullyQualifiedComponents.Add(tagHelper);
 
-                    StringSegment typeName = tagHelper.GetTypeName();
+                    if (currentNamespace is null)
+                    {
+                        continue;
+                    }
+
                     if (tagHelper.IsChildContentTagHelper())
                     {
                         // If this is a child content tag helper, we want to add it if it's original type is in scope.
                         // E.g, if the type name is `Test.MyComponent.ChildContent`, we want to add it if `Test.MyComponent` is in scope.
-                        TrySplitNamespaceAndType(typeName, out var namespaceTextSpan, out var _);
-                        typeName = GetTextSpanContent(namespaceTextSpan, typeName);
+                        TrySplitNamespaceAndType(tagHelper, out var typeNamespace);
```

Comment:
The diff in this PR are a bit of mess but might be better appreciated when viewed side-by-side. In essence, the change caches the split namespace / type name on TagHelperDescriptor and re-uses it. For this, we have to flow the descriptor around. The ChildComponent taghelper has some special parsing behavior which makes it difficult to cache, so we just parse it every time.

Reply:
Yep -- took a while to grok the diff but I see that most of the relevant changes are the layers of indirection to support caching when invoking `TrySplitNamespaceAndType`. Seems legit. 👍🏽